src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +144,211 @@ internal class DiagnosticComputer
+                    throw;
+                }
+                finally
+                {
+                    // Step 6:
+                    //  - Remove the 'cancellationTokenSource' for completed or cancelled task.
+                    //    For the case where the computeTask was cancelled, we will create a new
+                    //    'cancellationTokenSource' for the retry.
+                    lock (s_gate)
+                    {
+                        var removed = s_cancellationSourcesForNormalPriorityComputeTasks.Remove(cancellationTokenSource);
+                        Debug.Assert(removed);
+                    }
+                }
+            }
+
+            throw ExceptionUtilities.Unreachable();
+
+            Task<SerializableDiagnosticAnalysisResults> CreateComputeTask(CancellationTokenSource cancellationTokenSource, CancellationToken cancellationToken)
+            {
+                return Task.Run(async () =>
+                {
+                    // Create a linked cancellation source to allow high priority tasks to cancel normal priority tasks.
+                    var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationTokenSource.Token, cancellationToken);
+                    cancellationToken = linkedCancellationTokenSource.Token;
+
+                    try
+                    {
+                        return await GetDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, cancellationToken).ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        linkedCancellationTokenSource.Dispose();
+                    }
+                }, cancellationToken);
+            }
+
+            static async Task WaitForHighPriorityTasksAsync(CancellationToken cancellationToken)
+            {
+                // We loop continuously until we have an empty high priority task queue.
+                while (true)
+                {
+                    ImmutableArray<Task> highPriorityTasksToAwait;
+                    lock (s_gate)
+                    {
+                        highPriorityTasksToAwait = s_highPriorityComputeTasks.ToImmutableArrayOrEmpty();
+                    }
+
+                    if (highPriorityTasksToAwait.IsEmpty)
+                    {
+                        return;
+                    }
+
+                    foreach (var task in highPriorityTasksToAwait)
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        try
+                        {
+                            await task.ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            // Gracefully ignore cancellations for high priority tasks.
+                        }
```

Comment:
```suggestion
                        await task.NoThrowAwaitable().ConfigureAwait(false);
```

(or something like that).